src/lib/input/input-container.ts

```diff
@@ -311,9 +295,8 @@ export class MdInputDirective implements OnChanges, OnDestroy, DoCheck {
   /** Re-evaluates the error state. This is only relevant with @angular/forms. */
   private _updateErrorState() {
     const oldState = this._isErrorState;
-    const control = this._ngControl;
-    const parent = this._parentFormGroup || this._parentForm;
-    const newState = control && this.errorStateMatcher(control.control as FormControl, parent);
+    const errorMatcher = this.errorStateMatcher || this._errorOptions.errorStateMatcher;
```

Comment:
This is bad because it loses the error options context. Honestly, I think we should just get rid of `ErrorOptions` since there are currently no other options. and make `ErrorStateMatcher` a class rather than a function. This was we can easily support overriding via `@Input` like we want to do here and the stepper can easily provide its context-dependent matcher